tracks/python/exercises/palindrome-products/mentoring.md

```diff
@@ -0,0 +1,172 @@
+# Mentoring
+
+## Problem and challenges
+
+The problem asks for two functions, `smallest` and `largest`, that will respectively give the smallest and the largest palindrome number that is a product of at least one pair of factors in a given range.
+
+The main challenge of this problem is performance. A naive **O(N²)** solution can lead to timeouts when running the tests online.
+
+## Reasonable solution
+
+```python
+from typing import Callable, Optional, Sequence
+
+Factors = list[list[int]]
+Result = tuple[Optional[int], Factors]
+Comparison = Callable[[int, int], bool]
+
+
+def is_palindrome(num: int) -> bool:
+    textual = str(num)
+    return textual == textual[::-1]
+
+
+def search(factor_range: Sequence[int],
+           better_found: Comparison) -> Result:
+    factor_range = list(factor_range)
+    found = None
+    factors: Factors = []
+    for i in factor_range:
+        for j in factor_range:
+            num = i*j
+            # note: further values of j will yield even worse num
+            if found and not better_found(num, found):
+                break
+            if is_palindrome(num):
+                if num == found:
+                    factors.append([i, j])
+                else:
+                    factors = [[i, j]]
+                found = num
+    return (found, factors)
+
+
+def largest(min_factor: int, max_factor: int) -> Result:
+    if min_factor > max_factor:
+        raise ValueError('min must be <= max')
+    # decrementing for time optimization (see note above)
+    return search(range(max_factor, min_factor - 1, -1),
+                  lambda num, found: num >= found)
+
+
+def smallest(min_factor: int, max_factor: int) -> Result:
+    if min_factor > max_factor:
+        raise ValueError('min must be <= max')
+    return search(range(min_factor, max_factor + 1),
+                  lambda num, found: num <= found)
+```
+
+## Common suggestions
+
+- Dealing with correctness, suggest clearing the previous result if a better one is found.
+- Point out this observation for performance:
+    -  For two palindromes *p1=x\*y1* and *p2=x\*y2*, *p1>p2* if *y1>y2*.
+- Suggest avoiding code duplication between `largest` and `smallest`.
+- Complexity discusson is not trivial for this problem. Prefer informal wording such as *faster* or *slower*. 
+- Only for the rare mentee that has mastered the reasonable solution at first try and asked for more, steer them towards the faster solution below.
```

Comment:
Consider not classifying the mentee.  It may be a common mentee that has mastered the reasonable solution on first try.

On a first or second iteration, or where the author has asked for more... will redirect the attention to the effort, not on some imagined specialness of the author.